File: leetcode-cn/Python/112.路径总和.py
#
# @lc app=leetcode.cn id=112 lang=python3
#
# [112] 路径总和
#


# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        ans = False

        def help(root, target):
            if root is None:
                return

            nonlocal ans

            tmp = target - root.val

            if tmp == 0 and root.left is None and root.right is None:
                ans = True

            help(root.left, tmp)
            help(root.right, tmp)

        help(root, sum)

        return ans


# @lc code=end

File: leetcode-cn/C++/112.路径总和.cpp
/*
 * @lc app=leetcode.cn id=112 lang=cpp
 *
 * [112] 路径总和
 */

// @lc code=start
// struct TreeNode
// {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };

class Solution
{
public:
    bool hasPathSum(TreeNode *root, int targetSum)
    {
        if (root == nullptr)
        {
            return false;
        }

        targetSum = targetSum - root->val;

        if (targetSum == 0 && !root->left && !root->right)
        {
            return true;
        }

        return hasPathSum(root->left, targetSum) || hasPathSum(root->right, targetSum);
    }
};
// @lc code=end
